File: pt-1/3801.minimum-cost-to-merge-sorted-lists.java
/*
 * @lc app=leetcode id=3801 lang=java
 *
 * [3801] Minimum Cost to Merge Sorted Lists
 *
 * Bitmask DP over subsets of lists. Precompute each subset's total length and
 * merged median; then split every subset into two parts and pay both subcosts
 * plus merge length and median difference.
 *
 * Java note: subset medians are found by binary search over all distinct values;
 * Arrays.binarySearch-style upper bounds count elements <= candidate in each
 * sorted source list.
 *
 * Time: O(3^m + 2^m * m * VlogV) for m lists. Space: O(2^m).
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

// @lc code=start
class Solution {
    public long minMergeCost(int[][] lists) {
        int m = lists.length;
        int maskCount = 1 << m;
        int[] lengths = new int[maskCount];
        int[][] members = new int[maskCount][];

        members[0] = new int[0];
        for (int mask = 1; mask < maskCount; mask++) {
            int bit = mask & -mask;
            int index = Integer.numberOfTrailingZeros(bit);
            int prev = mask ^ bit;
            lengths[mask] = lengths[prev] + lists[index].length;
            members[mask] = Arrays.copyOf(members[prev], members[prev].length + 1);
            members[mask][members[mask].length - 1] = index;
        }

        int[] values = distinctValues(lists);
        int[][] countLeq = new int[m][values.length];
        for (int i = 0; i < m; i++) {
            for (int v = 0; v < values.length; v++) {
                countLeq[i][v] = upperBound(lists[i], values[v]);
            }
        }

        int[] medians = new int[maskCount];
        for (int mask = 1; mask < maskCount; mask++) {
            medians[mask] = subsetMedian(mask, lengths, members, values, countLeq);
        }

        long inf = Long.MAX_VALUE / 4;
        long[] dp = new long[maskCount];
        Arrays.fill(dp, inf);
        for (int i = 0; i < m; i++) {
            dp[1 << i] = 0;
        }

        for (int mask = 1; mask < maskCount; mask++) {
            if ((mask & (mask - 1)) == 0) {
                continue;
            }
            long best = inf;
            int sub = (mask - 1) & mask;
            while (sub > 0) {
                int other = mask ^ sub;
                if (sub < other) {
                    long candidate = dp[sub] + dp[other] + lengths[mask]
                            + Math.abs((long) medians[sub] - medians[other]);
                    best = Math.min(best, candidate);
                }
                sub = (sub - 1) & mask;
            }
            dp[mask] = best;
        }
        return dp[maskCount - 1];
    }

    private int subsetMedian(int mask, int[] lengths, int[][] members, int[] values, int[][] countLeq) {
        int target = (lengths[mask] - 1) / 2 + 1;
        int left = 0;
        int right = values.length - 1;
        while (left < right) {
            int mid = (left + right) >>> 1;
            int count = 0;
            for (int listIndex : members[mask]) {
                count += countLeq[listIndex][mid];
            }
            if (count >= target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return values[left];
    }

    private int[] distinctValues(int[][] lists) {
        List<Integer> all = new ArrayList<>();
        for (int[] list : lists) {
            for (int value : list) {
                all.add(value);
            }
        }
        int[] arr = new int[all.size()];
        for (int i = 0; i < all.size(); i++) {
            arr[i] = all.get(i);
        }
        Arrays.sort(arr);
        int unique = 0;
        for (int value : arr) {
            if (unique == 0 || arr[unique - 1] != value) {
                arr[unique++] = value;
            }
        }
        return Arrays.copyOf(arr, unique);
    }

    private int upperBound(int[] arr, int target) {
        int lo = 0;
        int hi = arr.length;
        while (lo < hi) {
            int mid = (lo + hi) >>> 1;
            if (arr[mid] <= target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
}
// @lc code=end

File: pt-1/3801.minimum-cost-to-merge-sorted-lists.go
package leetcode

//
// @lc app=leetcode id=3801 lang=golang
//
// [3801] Minimum Cost to Merge Sorted Lists
//
// Notes
// This is subset DP over lists. Precompute each subset's total length and
// median, then split every subset into two non-empty parts and pay child costs
// plus total length plus median distance. Sorted input lists allow median
// queries by binary searching over the global distinct values and precomputed
// per-list <= counts. Time: O(3^m + 2^m*m*log V). Space: O(2^m*m).
//
// @lc code=start

import "sort"

func MinMergeCost3801(lists [][]int) int {
	listCount := len(lists)
	maskCount := 1 << listCount

	lengths := make([]int, maskCount)
	medians := make([]int, maskCount)
	members := make([][]int, maskCount)
	for mask := 1; mask < maskCount; mask++ {
		lowestBit := mask & -mask
		listIndex := bitIndex3801(lowestBit)
		previousMask := mask ^ lowestBit
		lengths[mask] = lengths[previousMask] + len(lists[listIndex])
		members[mask] = append(append([]int(nil), members[previousMask]...), listIndex)
	}

	distinctValues := []int{}
	for _, values := range lists {
		distinctValues = append(distinctValues, values...)
	}
	sort.Ints(distinctValues)
	unique := distinctValues[:0]
	for _, value := range distinctValues {
		if len(unique) == 0 || unique[len(unique)-1] != value {
			unique = append(unique, value)
		}
	}
	distinctValues = unique

	countLeq := make([][]int, listCount)
	for i, values := range lists {
		countLeq[i] = make([]int, len(distinctValues))
		for j, candidate := range distinctValues {
			countLeq[i][j] = sort.Search(len(values), func(pos int) bool {
				return values[pos] > candidate
			})
		}
	}

	findSubsetMedian := func(mask int) int {
		target := (lengths[mask]-1)/2 + 1
		left, right := 0, len(distinctValues)-1
		for left < right {
			middle := (left + right) / 2
			count := 0
			for _, listIndex := range members[mask] {
				count += countLeq[listIndex][middle]
			}
			if count >= target {
				right = middle
			} else {
				left = middle + 1
			}
		}
		return distinctValues[left]
	}

	for mask := 1; mask < maskCount; mask++ {
		medians[mask] = findSubsetMedian(mask)
	}

	const inf = 1 << 60
	dp := make([]int, maskCount)
	for i := range dp {
		dp[i] = inf
	}
	for i := 0; i < listCount; i++ {
		dp[1<<i] = 0
	}

	for mask := 1; mask < maskCount; mask++ {
		if mask&(mask-1) == 0 {
			continue
		}
		best := inf
		totalLength := lengths[mask]
		for submask := (mask - 1) & mask; submask > 0; submask = (submask - 1) & mask {
			other := mask ^ submask
			if submask < other {
				candidate := dp[submask] + dp[other] + totalLength + abs3801(medians[submask]-medians[other])
				if candidate < best {
					best = candidate
				}
			}
		}
		dp[mask] = best
	}

	return dp[maskCount-1]
}

func bitIndex3801(bit int) int {
	index := 0
	for bit > 1 {
		bit >>= 1
		index++
	}
	return index
}

func abs3801(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// @lc code=end
